### ISSUE TITLE
Avoiding unneccessary duplication of asset dependencies

### ISSUE THREAD

**Issue by sbstnkhlr** [focustense/easymod#132]
As far as I understand it, EasyNPC creates a sort-of standalone Mod, incorporating all dependencies that are required to make up every NPC it touches. That means that al lot of assets from those dependency mods end up in the merged BSA, correct? 

In my case I see a whole lot of meshes and textures from KS Hairdo's in my BSA for example. 

This is nice if you want to have a standalone and portable Build of your NPC loadout, but it duplicates a lot of assets unneccesarily. 

Would it be possible to add a mode in which Easy NPC checks if the last "natural" conflict winner of a file is identical to the one it would incorporate into the BSA ? 

Of course, to make this sort of "reduced" Merge functional, the asset depency mods would have to be added as Masters for the merge.

TL;DR - EasyNPC copies assets from mods that are never conflicted in the first place or are final conflict winners anyway. That's nice to make a personal standalone NPC mod, but it potentially wastes a LOT of space.

**Comment by focustense** [comment]
Thanks for the feedback and the suggestion. I've given it some consideration and decided not to pursue it, for a few reasons:

1. Creating standalone mods is the stated purpose of EasyNPC. It's always going to be the case that some people find uses for a product that's different (sometimes very different) from its intended purpose, and that's great - however, adding this type of feature would lead to unpredictable and often unreliable behavior, confusing the majority of users and creating a major source of support headaches for me personally. As the project moves toward 1.0, the trend has been and should continue to be toward greater stability and fewer hazards.

2. It is not actually possible to do this in a reliable or consistent fashion. Shared resource mods may not even be in the load order - lots of mods use their own copies of the assets from KS Hairdos, High Poly Head, and so on. Second, there may not be an ESP/ESM at all for the shared mod, sometimes resources are simply provided as loose files. Third, even if they are in the load order, there is no algorithmic method to accurately determine the primary source because there is no metadata to describe it. And finally, if we could somehow determine the primary source, that might turn out to be one of the NPC overhauls that is supposed to be disabled at the end, which means it is incorrect behavior to add a master dependency.

3. Notwithstanding both [1] and [2] above, the benefits if implemented do not really outweigh the ongoing cost and complexity. The extra space adds up to perhaps 1-2 GB in a build where the original overhauls and resources would have taken at least 30-50 GB and where a "fully-loaded" _and optimized_ (BSA-packed) game build would be 80-100 GB, which is not a significant improvement. If the issue is one of staging space - that is, space required to complete the build (pre-BSA) as opposed to space used by the final build - then a feature such as [building directly to BSA](https://github.com/focustense/easymod/issues/73) has much greater utility. I could also potentially consider options such as separating the facegen data and shared-resource data into separate BSAs.

4. If there is an urgent need to reclaim the space, then it is not that hard in principle: extract the BSAs, delete everything that's not facegendata, repack the BSAs. You can use MO2 to see which files are conflict winners and thus, in theory, can be removed. This puts the onus on the person making these edits to get them right, which is how I'd want it to be, because it requires some risky assumptions and in general a human-level understanding of all the mods involved.

Perhaps in a post-1.0 version I might revisit some of the assumptions, but as of now it's not a good fit.

---

P.S. When global replacers start to be supported (#6, #8, etc.), then there will already be a system in place for mod-specific rules, and at that time, I _may_ decide to add a "don't copy assets" flag that could be opted into on a per-mod basis. That would give the power users _some_ level of control, while keeping the responsibility on them to manage it correctly. No timeline or guarantee for that, but it is a possibility.
